Not able to build with SYCL backend on Nvidia GPU anymore

Configuring Kokkos with the SYCL backend on an Nvidia GPU does not work anymore on `develop` branch, while it works for the 4.6.2 release:

```sh
cmake -B build \
  -DCMAKE_BUILD_TYPE=Release \
  -DKokkos_ENABLE_SYCL=ON \
  -DKokkos_ENABLE_UNSUPPORTED_ARCHS=ON \
  -DKokkos_ARCH_AMPERE86=ON \
  -DCMAKE_CXX_COMPILER=icpx
```

With CMake 3.28, Intel compiler 2025.0.0, on an A500 GPU with Cuda 3.28. Error message:

```log
CMake Error at cmake/kokkos_functions.cmake:987 (message):
  The compiler for CXX can not consume flag(s) -fsycl
  -fno-sycl-id-queries-fit-in-int -fsycl-dead-args-optimization
  -fsycl-unnamed-lambda -fno-sycl-rdc -fsycl-targets=nvptx64-nvidia-cuda
  -Xsycl-target-backend=nvptx64-nvidia-cuda --cuda-gpu-arch=sm_86 in
  combination with the CMAKE_CXX_FLAGS=.  Please check the given
  configuration.
Call Stack (most recent call first):
  cmake/kokkos_tribits.cmake:325 (kokkos_check_flags)
  cmake/kokkos_tribits.cmake:389 (kokkos_set_library_properties)
  core/src/CMakeLists.txt:136 (kokkos_add_library)
```

The bug appears after the merge of #7891.

In `cmake/CMakeFiles/CMakeConfigureLog.yaml`, I think I may have identified a new check induced by the aforementioned PR which fails and stops the configuration:

```yaml
   kind: "try_compile-v1"
    backtrace:
      - "/usr/share/cmake-3.28/Modules/Internal/CheckSourceCompiles.cmake:101 (try_compile)"
      - "/usr/share/cmake-3.28/Modules/Internal/CheckCompilerFlag.cmake:18 (cmake_check_source_compiles)"
      - "/usr/share/cmake-3.28/Modules/CheckCompilerFlag.cmake:52 (cmake_check_compiler_flag)"
      - "cmake/kokkos_functions.cmake:985 (check_compiler_flag)"
      - "cmake/kokkos_tribits.cmake:325 (kokkos_check_flags)"
      - "cmake/kokkos_tribits.cmake:389 (kokkos_set_library_properties)"
      - "core/src/CMakeLists.txt:136 (kokkos_add_library)"
    checks:
      - "Performing Test KOKKOS_COMPILE_OPTIONS_CHECK"
    directories:
      source: "/home/<<<REDACTED>>>/Projets/kokkos/build/sycl/ampere/intel/2025.0.0/release/CMakeFiles/CMakeScratch/TryCompile-PyusAU"
      binary: "/home/<<<REDACTED>>>/Projets/kokkos/build/sycl/ampere/intel/2025.0.0/release/CMakeFiles/CMakeScratch/TryCompile-PyusAU"
    cmakeVariables:
      CMAKE_CXX_FLAGS: ""
      CMAKE_CXX_FLAGS_DEBUG: "-g -Rno-debug-disables-optimization"
      CMAKE_EXE_LINKER_FLAGS: ""
      CMAKE_MODULE_PATH: "cmake/Modules;/home/<<<REDACTED>>>/Projets/kokkos/cmake/Modules/"
    buildResult:
      variable: "KOKKOS_COMPILE_OPTIONS_CHECK"
      cached: true
      stdout: |
        Change Dir: '/home/<<<REDACTED>>>/Projets/kokkos/build/sycl/ampere/intel/2025.0.0/release/CMakeFiles/CMakeScratch/TryCompile-PyusAU'
        
        Run Build Command(s): /usr/bin/cmake -E env VERBOSE=1 /usr/bin/gmake -f Makefile cmTC_907f2/fast
        /usr/bin/gmake  -f CMakeFiles/cmTC_907f2.dir/build.make CMakeFiles/cmTC_907f2.dir/build
        gmake[1]: Entering directory '/volatile/<<<REDACTED>>>/projects/kokkos/build/sycl/ampere/intel/2025.0.0/release/CMakeFiles/CMakeScratch/TryCompile-PyusAU'
        Building CXX object CMakeFiles/cmTC_907f2.dir/src.cxx.o
        /product/ubuntu24-x86_64/apps/intel-compilers/2025.0.0/compiler/2025.0/bin/icpx -DKOKKOS_COMPILE_OPTIONS_CHECK  -fsycl -fno-sycl-id-queries-fit-in-int -fsycl-dead-args-optimization -fsycl-unnamed-lambda -fno-sycl-rdc -std=gnu++17   -fsycl -fno-sycl-id-queries-fit-in-int -fsycl-dead-args-optimization -fsycl-unnamed-lambda -fno-sycl-rdc -fsycl-targets=nvptx64-nvidia-cuda -Xsycl-target-backend=nvptx64-nvidia-cuda --cuda-gpu-arch=sm_86 -MD -MT CMakeFiles/cmTC_907f2.dir/src.cxx.o -MF CMakeFiles/cmTC_907f2.dir/src.cxx.o.d -o CMakeFiles/cmTC_907f2.dir/src.cxx.o -c /home/<<<REDACTED>>>/Projets/kokkos/build/sycl/ampere/intel/2025.0.0/release/CMakeFiles/CMakeScratch/TryCompile-PyusAU/src.cxx
        Linking CXX executable cmTC_907f2
        /usr/bin/cmake -E cmake_link_script CMakeFiles/cmTC_907f2.dir/link.txt --verbose=1
        /product/ubuntu24-x86_64/apps/intel-compilers/2025.0.0/compiler/2025.0/bin/icpx  -fsycl -fno-sycl-id-queries-fit-in-int -fsycl-dead-args-optimization -fsycl-unnamed-lambda -fno-sycl-rdc CMakeFiles/cmTC_907f2.dir/src.cxx.o -o cmTC_907f2 
        icpx: error: linked binaries do not contain expected 'spir64-unknown-unknown' target; found targets: 'nvptx64-nvidia-cuda-sm_86', this is not supported with '-fno-sycl-rdc'
        gmake[1]: *** [CMakeFiles/cmTC_907f2.dir/build.make:100: cmTC_907f2] Error 1
        gmake[1]: Leaving directory '/volatile/<<<REDACTED>>>/projects/kokkos/build/sycl/ampere/intel/2025.0.0/release/CMakeFiles/CMakeScratch/TryCompile-PyusAU'
        gmake: *** [Makefile:127: cmTC_907f2/fast] Error 2
        
      exitCode: 2
```

The Jenkins CI job using SYCL on an A30 is still working fine, so I'm wondering if this is due to the GPU model?

--- a/cmake/kokkos_functions.cmake
+++ b/cmake/kokkos_functions.cmake
@@ -962,7 +962,7 @@ endfunction()
 #       FLAGS       --> flags to check
 #
 function(kokkos_check_flags)
-  cmake_parse_arguments(INP "COMPILER;LINKER" "LANGUAGE" "FLAGS" ${ARGN})
+  cmake_parse_arguments(INP "COMPILER;LINKER" "LANGUAGE" "FLAGS;LINKER_FLAGS" ${ARGN})
 
   # do nothing if no flags are given
   if(NOT INP_FLAGS)
@@ -982,6 +982,9 @@ function(kokkos_check_flags)
     include(CheckCompilerFlag)
     #delete cache so we always do the check
     unset(KOKKOS_COMPILE_OPTIONS_CHECK CACHE)
+    if(INP_LINKER_FLAGS)
+      set(CMAKE_REQUIRED_LINK_OPTIONS ${INP_LINKER_FLAGS})
+    endif()
     check_compiler_flag(${INP_LANGUAGE} "${QUOTED_FLAGS}" KOKKOS_COMPILE_OPTIONS_CHECK)
     if(NOT KOKKOS_COMPILE_OPTIONS_CHECK)
       message(

--- a/cmake/kokkos_functions.cmake
+++ b/cmake/kokkos_functions.cmake
@@ -983,7 +983,11 @@ function(kokkos_check_flags)
     #delete cache so we always do the check
     unset(KOKKOS_COMPILE_OPTIONS_CHECK CACHE)
     if(INP_LINKER_FLAGS)
-      set(CMAKE_REQUIRED_LINK_OPTIONS ${INP_LINKER_FLAGS})
+      #check_compiler/linker_flag requires a whitespace separated list
+      string(REPLACE ";" " " WHITESPACE_LINKER_FLAGS "${INP_LINKER_FLAGS}")
+      # icpx adds "-device ..." options that need quotes (which CMake removes). We need to add them here again.
+      string(REGEX REPLACE "(-device [A-Za-z0-9_\\\\.]*)" "\"\\1\"" QUOTED_LINKER_FLAGS "${WHITESPACE_LINKER_FLAGS}")
+      set(CMAKE_REQUIRED_LINK_OPTIONS "${QUOTED_LINKER_FLAGS}")
     endif()
     check_compiler_flag(${INP_LANGUAGE} "${QUOTED_FLAGS}" KOKKOS_COMPILE_OPTIONS_CHECK)
     if(NOT KOKKOS_COMPILE_OPTIONS_CHECK)

--- a/cmake/kokkos_functions.cmake
+++ b/cmake/kokkos_functions.cmake
@@ -983,11 +983,9 @@ function(kokkos_check_flags)
     #delete cache so we always do the check
     unset(KOKKOS_COMPILE_OPTIONS_CHECK CACHE)
     if(INP_LINKER_FLAGS)
-      #check_compiler/linker_flag requires a whitespace separated list
-      string(REPLACE ";" " " WHITESPACE_LINKER_FLAGS "${INP_LINKER_FLAGS}")
       # icpx adds "-device ..." options that need quotes (which CMake removes). We need to add them here again.
-      string(REGEX REPLACE "(-device [A-Za-z0-9_\\\\.]*)" "\"\\1\"" QUOTED_LINKER_FLAGS "${WHITESPACE_LINKER_FLAGS}")
-      set(CMAKE_REQUIRED_LINK_OPTIONS ${QUOTED_LINKER_FLAGS})
+      string(REGEX REPLACE "(-device [A-Za-z0-9_\\\\.]*)" "\"\\1\"" QUOTED_LINKER_FLAGS "${INP_LINKER_FLAGS}")
+      set(CMAKE_REQUIRED_LINK_OPTIONS "${QUOTED_LINKER_FLAGS}")
     endif()
     check_compiler_flag(${INP_LANGUAGE} "${QUOTED_FLAGS}" KOKKOS_COMPILE_OPTIONS_CHECK)
     if(NOT KOKKOS_COMPILE_OPTIONS_CHECK)

--- a/cmake/kokkos_functions.cmake
+++ b/cmake/kokkos_functions.cmake
@@ -987,7 +987,7 @@ function(kokkos_check_flags)
       string(REPLACE ";" " " WHITESPACE_LINKER_FLAGS "${INP_LINKER_FLAGS}")
       # icpx adds "-device ..." options that need quotes (which CMake removes). We need to add them here again.
       string(REGEX REPLACE "(-device [A-Za-z0-9_\\\\.]*)" "\"\\1\"" QUOTED_LINKER_FLAGS "${WHITESPACE_LINKER_FLAGS}")
-      set(CMAKE_REQUIRED_LINK_OPTIONS "${QUOTED_LINKER_FLAGS}")
+      set(CMAKE_REQUIRED_LINK_OPTIONS ${QUOTED_LINKER_FLAGS})
     endif()
     check_compiler_flag(${INP_LANGUAGE} "${QUOTED_FLAGS}" KOKKOS_COMPILE_OPTIONS_CHECK)
     if(NOT KOKKOS_COMPILE_OPTIONS_CHECK)

--- a/cmake/kokkos_tribits.cmake
+++ b/cmake/kokkos_tribits.cmake
@@ -322,7 +322,7 @@ function(KOKKOS_SET_LIBRARY_PROPERTIES LIBRARY_NAME)
        AND (NOT (KOKKOS_ENABLE_CUDA) OR ("${CMAKE_CXX_COMPILER}" MATCHES "nvcc_wrapper") OR (${KOKKOS_CXX_COMPILER_ID}
                                                                                              STREQUAL Clang))
     )
-      kokkos_check_flags(COMPILER LANGUAGE ${KOKKOS_COMPILE_LANGUAGE} FLAGS ${ALL_KOKKOS_COMPILER_FLAGS})
+      kokkos_check_flags(COMPILER LANGUAGE ${KOKKOS_COMPILE_LANGUAGE} FLAGS ${ALL_KOKKOS_COMPILER_FLAGS} LINKER_FLAGS ${KOKKOS_LINK_OPTIONS})
     endif()
   endif()
 

--- a/cmake/kokkos_tribits.cmake
+++ b/cmake/kokkos_tribits.cmake
@@ -322,7 +322,15 @@ function(KOKKOS_SET_LIBRARY_PROPERTIES LIBRARY_NAME)
        AND (NOT (KOKKOS_ENABLE_CUDA) OR ("${CMAKE_CXX_COMPILER}" MATCHES "nvcc_wrapper") OR (${KOKKOS_CXX_COMPILER_ID}
                                                                                              STREQUAL Clang))
     )
-      kokkos_check_flags(COMPILER LANGUAGE ${KOKKOS_COMPILE_LANGUAGE} FLAGS ${ALL_KOKKOS_COMPILER_FLAGS} LINKER_FLAGS ${KOKKOS_LINK_OPTIONS})
+      kokkos_check_flags(
+        COMPILER
+        LANGUAGE
+        ${KOKKOS_COMPILE_LANGUAGE}
+        FLAGS
+        ${ALL_KOKKOS_COMPILER_FLAGS}
+        LINKER_FLAGS
+        ${KOKKOS_LINK_OPTIONS}
+      )
     endif()
   endif()
 

